### PR TITLE
#2222 Create a Base Class for CsvSink Tests

### PR DESCRIPTION
**Pull request by eechih**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #2222 )
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-connector/src/test/scala/com/island/ohara/connector/CsvSinkTestBase.scala [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+  private val TOPIC = TopicKey.of(CommonUtils.randomString(5), CommonUtils.randomString(5))
+  private val DATA = Row.of(Cell.of("a", "abc"), Cell.of("b", 123), Cell.of("c", true))
+  private val DATA_COUNT = 1000
+  private val SCHEMA: Seq[Column] = Seq(
+    Column.builder().name("a").dataType(DataType.STRING).order(1).build(),
+    Column.builder().name("b").dataType(DataType.INT).order(2).build(),
+    Column.builder().name("c").dataType(DataType.BOOLEAN).order(3).build()
+  )
+
+  @BeforeClass
```
[the remark this answers is not in the document]
Comment (eechih) [reply ohara-connector/src/test/scala/com/island/ohara/connector/CsvSinkTestBase.scala]: 就我目前的理解，每個 Sink Connector 都有“自己的” Offset（消費記錄），互不影響。
因此，同時有多個 Sink Connector 消費同一個 Topic 也不會有問題。

[the remark this answers is not in the document]
Comment (chia7712) [reply ohara-connector/src/test/scala/com/island/ohara/connector/CsvSinkTestBase.scala]: My question was “does test case see the data generated by other test cases”. Even if each connector has its offset, it is possible to see the data of other test case if the offset is “smaller” than the data offset. And the offset of new connector, of course, is larger than previous data by default, hence the scenario in my question never happen.

[the remark this answers is not in the document]
Comment (eechih) [reply ohara-connector/src/test/scala/com/island/ohara/connector/CsvSinkTestBase.scala]: got it, I will address it.

**Comment by eechih** [comment]
@chia7712 address your comments, please review again, thanks.

**Comment by eechih** [comment]
retry

**Comment by eechih** [comment]
retry

**Comment by eechih** [comment]
@chia7712 I've updated the PR, could you give it a look again?

File: ohara-connector/src/test/scala/com/island/ohara/connector/CsvSinkTestBase.scala [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+import com.island.ohara.kafka.Producer
+import com.island.ohara.testing.With3Brokers3Workers
+import org.junit.{After, Before, Test}
+import org.scalatest.Matchers
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+
+object CsvSinkTestBase extends With3Brokers3Workers with Matchers {
```
Comment (chia7712) [inline ohara-connector/src/test/scala/com/island/ohara/connector/CsvSinkTestBase.scala]: the objects (zk, bk and wk) in With3Brokers3Workers is (global) static member so you don't need to force scala to create static stuff for you. Hence, it is fine to move all fields from object scope to class so as to simplify test code.

Comment (eechih) [reply ohara-connector/src/test/scala/com/island/ohara/connector/CsvSinkTestBase.scala]: got it :)

**Comment by eechih** [comment]
> This PR may break API compatibility!!! ping @konekoya @wu87988622 see https://builds.is-land.com.tw/job/PreCommit-OHARA/5535/ for logs

@konekoya @wu87988622 Don't cate about this because I stopped Jenkins task manually.

**Comment by eechih** [comment]
Thanks for the review. @chia7712

**Pull request merged**